src/main/java/com/io/image/manager/service/ImageServiceImpl.java

```diff
@@ -92,27 +101,37 @@ public CacheResult fetchAndCacheImage(
         throw new ImageNotFoundException("Image not found at origin: " + origin.getUrl());
     }
 
-    private Optional<BufferedImage> fetchRemoteImage(OriginServer origin, String filename) {
-        try {
+    private Optional<BufferedImage> fetchRemoteImage(OriginServer origin, String filename, CloseableHttpClient client) {
             // FIXME: this does not look pretty
-            URL url = new URL(origin.getUrl() + filename);
-            URLConnection conn = url.openConnection();
-            Map<String,List<String>> headers = conn.getHeaderFields();
-            Optional<List<String>> cacheHead = Optional.empty();
-            if(headers.containsKey("Cache-Control")) {
-                cacheHead = Optional.of(headers.get("Cache-Control"));
-            }
-            Optional<List<String>> eTag = Optional.empty();
-            if(headers.containsKey("ETag")) {
-                 eTag = Optional.of(headers.get("ETag"));
-            }
-            byte[] imgBytes = conn.getInputStream().readAllBytes();
-            BufferedImage image = ImageIO.read(new ByteArrayInputStream(imgBytes));
-            if (image == null) {
+            try {
+                    HttpGet get = new HttpGet(origin.getUrl() + filename);
+                    CloseableHttpResponse response = client.execute(get);
+                    try{
+                    HttpEntity entity = response.getEntity();
+                    InputStream is = entity.getContent();
+
+                    Optional<String> cacheHead = Optional.empty();
+                    if (response.getHeaders("Cache-Control") != null) {
+                        cacheHead = Optional.of(response.getHeaders("Cache-Control").toString());
+                    }
+                    Optional<String> eTag = Optional.empty();
+                    if (response.getHeaders("ETag") != null) {
```

Comment:
Can you please check if those headers are case sensitive? We can fail due to case sensitivity really easily. If it is not case senstitive then change it all to lower case.

Reply:
Those headers are case sensitive, I am not sure how apache hadnles case sensitivity, but it is probably safer to leave them as they are.

Reply:
Well, actually acording to RFC headers are case insensitive, still in my opinion it is safer to leave them in this form